File: build-all.py
import os
import subprocess
import shutil
import sys
import json

OUTPUT_DIR = "dist_all"

try:
    base_url = sys.argv[1]
except IndexError:
    base_url = ""

if not base_url:
    print("Base URL (first arg) cannot be empty")
    exit(1)

def is_app(path):
    return os.path.isdir(path) and 'vite.config.ts' in os.listdir(path)

def already_installed(app):
    return app in os.listdir(OUTPUT_DIR)

def app_url(app):
    return os.path.join(base_url, app)

def print_app_info_markdown(app):
    with open("./{}/package.json".format(app), "r") as f:
        data = json.load(f)

    name = data.get('name', app)
    desc = data.get('description', '')
    url = app_url(app)
    print("{} ([Demo]({}))".format(name, url))
    print(desc)
    print()
    print("TODO: Write descriptions on all applications")
    print("TODO: (HARD to implement) create a software that generates screenshots (inside a server using Docker with Puppeetter or something like that)")
    # TODO: Maybe one way to implement this is to create a microservice or a command (if the
    #       command is just one run of a self-removing docker container, then that'd be enough
    #       and compact... no need to open ports). Which gives the screenshot of a website.
    #       the screenshot shooting doesn't need to be in this script, it can be in a different
    #       deployment stage.

apps = [f for f in os.listdir() if is_app(f)]

if not os.path.exists(OUTPUT_DIR):
    os.mkdir(OUTPUT_DIR)

for app in apps:
    print("#################### Installing", app)
    if already_installed(app):
        print("Skipped")
        continue
    subprocess.call(["npm", "install"], cwd=app)
    subprocess.call(["npx", "tsc"], cwd=app)
    subprocess.call(["npx", "vite", "build", "--base={}".format(app_url(app))], cwd=app)
    source = "./{}/dist".format(app)
    dest = "./{}/{}".format(OUTPUT_DIR, app)
    shutil.move(source, dest, copy_function=shutil.copytree)

print()
print("Clean commands (execute manually):")
for app in apps:
    print("rm -rf ./{}/node_modules".format(app))

print()
print("Markdown information:")
for app in apps:
    print_app_info_markdown(app)

File: build-all.js
// TODO: Lint & format
const fs = require('fs-extra');
const path = require('path');
const child_process = require('child_process');
const { requireBaseUrlEnv, getAllAppNames, appUrl } = require('./util');

const OUTPUT_DIR = "dist_all";
const BASE_URL = requireBaseUrlEnv()

console.log(`Base URL: ${BASE_URL}`)

function alreadyInstalled(app) {
  return fs.existsSync(path.join(OUTPUT_DIR, app))
}

async function execute(cmd, args, cwd) {
  const opts = {
    stdio: "inherit",
    shell: true,
    cwd
  }

  return new Promise(resolve => {
    child_process.spawn(cmd, args, opts).on('close', resolve)
  })
}

async function main() {
  const apps = getAllAppNames()
  
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR)
  }
  
  for (const app of apps) {
    console.log(`#################### Installing ${app}`);
    
    if (alreadyInstalled(app)) {
      console.log("Skipped (folder exists)");
      continue;
    }
    
    const outDist = path.join(app, 'dist');

    if (fs.existsSync(outDist)) {
      fs.removeSync(outDist)
    }
  
    await execute('npm', ['install'], app)
    const compileResult = await execute('npx', ['tsc'], app)
    
    if (compileResult !== 0) {
      throw new Error(`(${app}) Compilation failed`)
    }

    await execute('npx', ['vite', 'build', `--base=${appUrl(app)}`], app)

    const cpDest = path.join(OUTPUT_DIR, app);
    await fs.cpSync(outDist, cpDest, { recursive: true })
  }
  
  console.log();
  console.log("Clean commands (execute manually):");
  for (const app of apps) {
    console.log(`rm -rf ./${app}/node_modules`);
  }
}

main().catch(console.error)
